Validate custom spend limit

Gudahtt opened the pull request:
The custom spend limit was previously not validated. It did have a minimum of zero set, but this didn't have any affect (that minimum is used for form constraint validation, and this field wasn't in a form).
The field was never checked to ensure the contents didn't exceed the maximum.

The field is now checked for values that exceed the maximum, and invalid values in general (including negative values).

The parameters to the `showEditApprovalPermissionModal` were also alphabetized to make them easier to read. In the course of doing this, I noticed that the origin was missing from one of the calls. This was responsible for the modal saying "Spend limit requested by undefined" when clicking "Edit" under the transaction details. This has been fixed.

Gudahtt committed:
Validate custom spend limit

The custom spend limit was previously not validated. It did have a
minimum of zero set, but this didn't have any affect (that minimum is
used for form constraint validation, and this field wasn't in a form).
The field was never checked to ensure the contents didn't exceed the
maximum.

The field is now checked for values that exceed the maximum, and
invalid values in general (including negative values).

The parameters to the `showEditApprovalPermissionModal` were also
alphabetized to make them easier to read. In the course of doing this,
I noticed that the origin was missing from one of the calls. This was
responsible for the modal saying "Spend limit requested by undefined"
when clicking "Edit" under the transaction details. This has been
fixed.

Gudahtt commented on ui/app/components/app/modals/edit-approval-permission/edit-approval-permission.component.js:
```diff
@@ -1,12 +1,18 @@
 import TextField from '../../../ui/text-field'
+import { calcTokenAmount } from '../../../../helpers/utils/token-util'
 import classnames from 'classnames'
+const BigNumber = require('bignumber.js')
```
This should use `import` syntax

Gudahtt commented on ui/app/components/app/modals/edit-approval-permission/edit-approval-permission.component.js:
```diff
@@ -155,10 +161,44 @@ export default class EditApprovalPermission extends PureComponent {
+      return t('spendLimitInvalid')
+    }
+
+    const maxTokenAmount = calcTokenAmount(MAX_UNSIGNED_256_INT, decimals)
```
This is not working correctly for some reason 😕 The limit is different than I'd expect. I'll need to investigate further.

Gudahtt replied:
Oh right, the number is too big, of course 🤦‍♂️. Need to use bignumber here as well.

Gudahtt commented:
This should be ready to review again now

rickycodes commented on ui/app/components/app/modals/edit-approval-permission/edit-approval-permission.component.js:
```diff
@@ -156,10 +161,44 @@ export default class EditApprovalPermission extends PureComponent {
     const { selectedOptionIsUnlimited, customSpendLimit } = this.state
+
+    const error = this.validateSpendLimit()
+    const disabled = Boolean(
```
I appreciate this logic getting moved out from the JSX

Gudahtt replied:
Yeah it can get pretty cluttered in there. I usually try to limit JSX-embedded logic to just a single ternary, and even then it's questionable.

danjm commented on ui/app/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:
```diff
@@ -209,10 +219,12 @@ export default class ConfirmApproveContent extends Component {
                     onEditClick: () => showEditApprovalPermissionModal({
                       customTokenAmount,
+                      decimals,
+                      origin,
```
With this fix, maybe we should also make `origin` required in the `editApprovalPermissions` component?

Gudahtt replied:
Yes, that's an excellent point, I will do that. I hadn't considered this because I fixed this in the first place accidentally 😅 then noticed looking back at the diff.

Gudahtt replied:
Fixed here: https://github.com/MetaMask/metamask-extension/pull/7920/commits/b723802f3ec71dd69fb4e341da64d89c186ce319

danjm commented on ui/app/components/app/modals/edit-approval-permission/edit-approval-permission.component.js:
```diff
@@ -156,10 +161,44 @@ export default class EditApprovalPermission extends PureComponent {
   }
 
+  validateSpendLimit () {
+    const { t } = this.context
```
This function looks sound and correct

danjm reviewed (dismissed):
Left a comment about a possible improvement, but it is not critical. This PR looks good.

Gudahtt committed:
Make `origin` a required prop of `edit-approval-permission`

Without this prop, the user is told that the spend limit was requested
by "undefined".

SvenMeyer commented:
I the "Spend limit" per transaction ?
... or the maximum a website can ever spend in total ?
